shared-libs/rules-engine/src/task-states.js

```diff
@@ -31,6 +45,25 @@ const States = {
   Failed: 'Failed',
 };
 
+const getDisplayWindow = (taskEmission) => {
+  const hasExistingDisplayWindow = taskEmission.startTime || taskEmission.endTime;
+  if (hasExistingDisplayWindow) {
+    return {
+      dueDate: taskEmission.dueDate,
+      startTime: taskEmission.startTime,
+      endTime: taskEmission.endTime,
+    };
+  }
+
+  const dueDate = moment(taskEmission.date);
+  return {
+    dueDate: dueDate.valueOf(),
+    startTime: dueDate.clone().startOf('day').subtract(taskEmission.readyStart || 0, 'days').valueOf(),
+    endTime: dueDate.clone().startOf('day').add(taskEmission.readyEnd || 0, 'days').add(1, 'day').valueOf() - 1,
```

Comment:
1. There's too much going on inline here and there's some duplication in date editing.
2. You're doing two additions to get the endTime which is unnecessary.
3. You don't need to clone the date as they are immutable - calling `add` returns a new moment object.
4. Should startTime and endTime be startDate and endDate respectively?
5. The `-1` at the end of the endTime calculation is probably unnecessary - that'll make it disappear 1ms earlier, right?

What about...

```js
const dueDate = moment(taskEmission.date).startOf('day');
const readyStart = taskEmission.readyStart || 0;
const readyEnd = (taskEmission.readyEnd || 0) + 1;
return {
  dueDate: dueDate.valueOf(),
  startDate: dueDate.subtract(readyStart, 'days').valueOf(),
  endDate: dueDate.add(readyEnd, 'days').valueOf(),
};
```

Reply:
Moments are not immutable and I believe the `clone` is necessary. https://momentjs.com/docs/#/manipulating/add/
> Mutates the original moment by adding time.

Reply:
> The -1 at the end of the endTime calculation is probably unnecessary - that'll make it disappear 1ms earlier, right?

If you take `moment().startOf('day').add(1, 'day')` you get a timestamp in tomorrow. 
If you subtract 1ms, you get a timestamp in today (11:59 pm).

Reply:
> Should startTime and endTime be startDate and endDate respectively?

Although this would be a decent description of the intent, I'd rather name the variable startTime since it is a timestamp.

Reply:
okay ya. i'll make that change

Reply:
> Moments are mutable

Wow, sorry, you're right, my memory failed me :(